enclave-runtime/src/tls_ra/tls_ra_client.rs

```diff
@@ -0,0 +1,203 @@
+		TlsClient { tls_stream, seal_handler, shard }
+	}
+
+	/// Read all data sent by the server of the specific shard.
+	///
+	/// We trust here that the server sends us the correct data, as
+	/// we do not have any way to test it.
+	fn read_shard(&mut self) -> EnclaveResult<()> {
+		self.write_shard()?;
+		self.read_all()
+	}
+
+	/// Send the shard of the state we want to receive to the provisioning server.
+	fn write_shard(&mut self) -> EnclaveResult<()> {
+		self.tls_stream.write(self.shard.as_bytes())?;
+		Ok(())
+	}
+
+	/// Read all relevant data sent by the server.
+	fn read_all(&mut self) -> EnclaveResult<()> {
+		let mut continue_reading = true;
+		while continue_reading {
+			continue_reading = self.read_and_seal()?;
+		}
+		info!("Successfully read all data sent by the state provisioning server.");
+		Ok(())
+	}
+
+	/// Read a server header / payload pair and directly seal the received data.
+	fn read_and_seal(&mut self) -> EnclaveResult<bool> {
+		let mut start_byte = [0u8; 1];
+		let read_size = self.tls_stream.read(&mut start_byte)?;
+		// If we're reading but there's no data: EOF.
+		if read_size == 0 {
+			return Ok(false)
+		}
+		let header = self.read_header(start_byte.to_vec())?;
+		let bytes = self.read_until(header.payload_length as usize)?;
+		match header.opcode {
+			Opcode::ShieldingKey => self.seal_handler.seal_shielding_key(&bytes)?,
+			Opcode::SigningKey => self.seal_handler.seal_signing_key(&bytes)?,
+			Opcode::State => self.seal_handler.seal_state(&bytes, &self.shard)?,
+		};
+		Ok(true)
+	}
+
+	/// Reads the payload header, indicating the sent payload length and type.
+	fn read_header(&mut self, start_bytes: Vec<u8>) -> EnclaveResult<TcpHeader> {
+		// The first sent byte indicates the payload type.
+		let opcode: Opcode = start_bytes[0].into();
+		// The 8 bytes following afterwards indicate the payload length.
+		let mut length_buffer = [0u8; 8];
+		self.tls_stream.read(&mut length_buffer)?;
+		let payload_length = u64::from_be_bytes(length_buffer);
+		debug!("payload_length: {}", payload_length);
+
+		Ok(TcpHeader::new(opcode, payload_length))
+	}
+
+	/// Read all bytes into a buffer of given length.
+	fn read_until(&mut self, length: usize) -> EnclaveResult<Vec<u8>> {
+		let mut bytes = vec![0u8; length];
+		self.tls_stream.read(&mut bytes)?;
+		Ok(bytes)
+	}
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn request_state_provisioning(
+	socket_fd: c_int,
+	sign_type: sgx_quote_sign_type_t,
+	shard: *const u8,
+	shard_size: u32,
+	skip_ra: c_int,
+) -> sgx_status_t {
+	let _ = backtrace::enable_backtrace("enclave.signed.so", PrintFormat::Short);
+	let shard = ShardIdentifier::from_slice(slice::from_raw_parts(shard, shard_size as usize));
+
+	let state_handler = Arc::new(GlobalFileStateHandler);
+	let seal_handler = SealHandler::<Rsa3072Seal, AesSeal, _>::new(state_handler);
+
+	if let Err(e) =
+		request_state_provisioning_internal(socket_fd, sign_type, shard, skip_ra, seal_handler)
+	{
+		return e.into()
```

Comment:
We should probably log the error here? This is something I discovered now when debugging the sidechain: The `into()` `sgx_status_t` loses all the error information, and we just get an `SGX_ERROR_UNEXPECTED`, which is not helpful. Logging `error!()` can mitigate that.